kod/util/gameevent/invasion/jonasatt.kod

```diff
@@ -0,0 +1,298 @@
+% Meridian 59, Copyright 1994-2012 Andrew Kirmse and Chris Kirmse.
+% All rights reserved.
+%
+% This software is distributed under a license that is described in
+% the LICENSE file that accompanies it.
+%
+% Meridian is a registered trademark.
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+JonasAttack is Invasion
+
+constants:
+
+   include blakston.khd
+
+resources:
+
+   JonasAttack_start_msg = \
+      "~B~k[Event]~n ~B~v The rebels are advancing!"
+
+   JonasAttack_end_msg = \
+      "~B~k[Event]~n ~B~v The rivals fall back in failure."
+
+   JonasAttack_phase2_msg = \
+      "~B~k[Event]~n ~B~v The rebels have brought mountain beast to assist them!"
+
+   JonasAttack_phase3_msg = \
+      "~B~k[Event]~n ~B~v The royal princess has sent her troops to assist in the attack."
+
+   JonasAttack_finalphase_msg = \
+      "~B~k[Event]~n ~B~v The royal guards have brought beast from the mountains to "
+      "assist them in there final attack!"
+
+classvars:
+
+   vrStartMessage = JonasAttack_start_msg
+   vrEndMessage = JonasAttack_end_msg
+
+   vbAnnounce = TRUE
+
+properties:
+
+   plEventRooms = $
+   piRebelTroop = 15
+   piTroll = 10
+   piPrincessTroop = 15
+   piGroundWormQueen = 10
+   piActiveRoom = RID_TOS, RID_TOS_EAST, RID_BLACKSTONE
+   piPhase = 0
+
+messages:
+
+   Constructor()
+   {
+      plEventRooms = [ RID_TOS, RID_TOS_EAST, RID_BLACKSTONE ];
+      if piActiveRoom = RID_TOS
+      {
+         plSpawnRects = [ [[54,5],[61,13]], [[49,11],[62,27]],
+                          [[54,29],[59,40]], [[14,16],[49,21]] ];
+      }
+      
+      if piActiveRoom = RID_BLACKSTONE
+      {
+         plSpawnRects = [ [[15,21],[15,38]], [[30,39],[30,25]],
+                             [[24,19],[24,37]], [[20,29],[39,29]] ];
+      }
+      
+      if piActiveRoom = RID_TOS_EAST
+      {
+         plSpawnRects = [ [[27,10],[13,5]], [[36,7],[42,16]],
+                          [[45,30],[37,25]], [[45,41],[23,22]] ];
+      }
+
+      propagate;
+   }
```

Comment:
Since you will be spawning monsters in up to 3 rooms this will need to be called more than just at the beginning, so name it something other than Constructor, You technically don't need a constructor at all.  Also change the propagate at the end to return, the parent class will not have a version of what you changed the name to.  I suggest this is where you put ChangeRoom() because every time you change the room the plSpawnRects will have to change to match that room.  You can put the room picking code right at the beginning of this message.